[Concurrency][Windows] Add Win32 native executors.

al45tair opened the pull request:
Added `Win32EventLoopExecutor` and `Win32ThreadPoolExecutor` as well as the `Win32NativeExecutorFactory`. Also added more executor tests, and fixed a bug in `CooperativeExecutor` that showed up in the test.

rdar://137111020

al45tair committed:
[Concurrency][Windows] Add Win32 native executors.

Added `Win32EventLoopExecutor` and `Win32ThreadPoolExecutor` as well as
the `Win32NativeExecutorFactory`.

rdar://137111020

al45tair committed:
[Concurrency] Add an executor test fixture and use it.

This also picked up a subtle problem with the `CooperativeExecutor`,
wherein it would run jobs out of order when they had equal priorities.
Fix this by adding a sequence number (and do the same fix to the Win32
native executor also, as it has the same problem).

rdar://137111020

al45tair committed:
[Concurrency][Win32] Minor tweaks for the Win32 executors.

A couple of little tweaks after the tests were added.

rdar://137111020

al45tair commented:
(First commit is from https://github.com/swiftlang/swift/pull/80636)

al45tair commented:
@swift-ci Please test

al45tair commented:
Linux WASI failure is just because I'm being too tight with the timing for the test.

al45tair committed:
[Concurrency] Tweak executor test, add missing `public`.

`Win32TaskExecutor` is missing a couple of `public`s; also, tweak the
`ExecutorFixture` to be more tolerant of delays.

rdar://137111020

al45tair commented:
Windows failure was some missing `public`s.

al45tair commented:
@swift-ci Please smoke test

ktoso commented on stdlib/public/Concurrency/CooperativeExecutor.swift:
```diff
@@ -70,6 +70,21 @@ extension ExecutorJob {
     }
   }
 
+  fileprivate var cooperativeExecutorSequence: UInt {
```
Maybe Sequence "Number"? I see the comment below that explains how we use it, maybe just the name could include number here

```suggestion
  fileprivate var cooperativeExecutorSequenceNr: UInt {
```

al45tair replied:
IMO `Sequence` is a clear enough name on its own — it can't really be anything other than a sequence number, since it's a variable that holds a number, and there _is_ a comment explaining what it's for. Plus if I put `Nr` on the end, someone else will tell me it needs to be `Number` and it's already a pretty long name…

ktoso replied:
Meh, okey :) Let's stick with this, not a big deal, it is private anyway 👍

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -303,15 +303,26 @@ public struct ExecutorJob: Sendable, ~Copyable {
+      }
+      return JobPriority(rawValue: raw)
+    }
+    set {
```
Hmmm are we really ok with making that settable by end users...? It would not cause priority escalation of child tasks if someone just randomly sets these like that etc Seems problematic?

ktoso replied:
Chatted offline that it's for testing, so maybe we can SPI it

al45tair committed:
[Concurrency] Tidy things up a bit.

Make the setter for `priority` SPI.

Remove `hThread` from the `Win32EventLoopExecutor` as we aren't using
it.

Improve the comment in `ExecutorFixture` explaining `maxDelayTolerance`.

Add warnings to `ExecutorFixture` if we exceed 125% of the requested
delay, but don't exceed the computed tolerance.

rdar://137111020

al45tair commented:
@swift-ci please smoke test Windows platform

al45tair commented:
@swift-ci please test

al45tair committed:
[Concurrency] Fix a couple of Win32EventLoopExecutor issues.

Fix running after executor is stopped, and add extra testing to the
`ExecutorFixture` to explicitly test that case.

Improve timer behaviour; we should target the fire time, not the
deadline.

rdar://137111020

al45tair committed:
[Concurrency][Tests] Increase tolerance for executor scheduling.

We've now got this set at 1/4s minimum.

rdar://137111020

al45tair commented:
@swift-ci Please smoke test

al45tair commented on stdlib/public/Concurrency/Win32Executor.swift:
```diff
@@ -0,0 +1,720 @@
+  }
+
+  deinit {
+    unsafe CloseHandle(hEvent!)
```
If we don't have an event object, that's definitely a fatal problem, so forced unwrap is fine here. Maybe we should check and `fatalError` on line 232.

al45tair commented on stdlib/public/Concurrency/Win32Executor.swift:
```diff
@@ -0,0 +1,720 @@
+@safe
+public final class Win32ThreadPoolExecutor: TaskExecutor, @unchecked Sendable {
+
+  private var cbeHighPriority = unsafe TP_CALLBACK_ENVIRON()
+  private var cbeLowPriority = unsafe TP_CALLBACK_ENVIRON()
+  private var cbeNormalPriority = unsafe TP_CALLBACK_ENVIRON()
```
I'm not inclined to change that in this PR, but we can do that in future if you think it's an improvement.

al45tair commented on stdlib/public/Concurrency/Win32Executor.swift:
```diff
@@ -0,0 +1,720 @@
+  }
+
+  fileprivate var win32TimestampIsIndirect: Bool {
+    return unsafe MemoryLayout<OpaquePointer>.stride * 2
```
I'll update the comments.

al45tair commented on test/Concurrency/Runtime/Inputs/ExecutorFixture.swift:
```diff
@@ -0,0 +1,432 @@
+  static func test(executor: some Executor) async -> Bool {
+    var result = true
+    print("Testing \(executor)")
+    if let runLoopExecutor = executor as? any RunLoopExecutor {
+      result = await test(runLoopExecutor: runLoopExecutor) && result
+    }
+    if let serialExecutor = executor as? any SerialExecutor {
+      result = await test(serialExecutor: serialExecutor) && result
+    }
+    #if !$Embedded && !SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY
+    if let schedulableExecutor = executor as? any SchedulableExecutor {
+      result = await test(schedulableExecutor: schedulableExecutor) && result
+    }
+    #endif
+    if let taskExecutor = executor as? any TaskExecutor {
+      result = await test(taskExecutor: taskExecutor) && result
+    }
```
You'd have to use `fallthrough` as well. I'm not sure it's really any better.

al45tair committed:
[Concurrency] Tweaks from review.

Minor changes from recent review.

rdar://137111020

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
Looks like there's something wrong with the `lit.cfg` for Windows; if I'm reading it right, it's _supposed_to be passing `-sdk` as part of the `%host-build-swift`, but it clearly isn't doing that, so it's been looking in the wrong place for modules.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
This is super annoying. The Windows tests are subtly broken — if a test even indirectly references the WinSDK module, it will fail because it tries to use the copy from the host toolchain instead of using the newly built copy. The Concurrency code is all done and working properly — this is just _another_ bug in the test system. Trying to figure out how to fix it at the moment.

al45tair committed:
[Tests][Windows] Pass the extra options through lit to make WinSDK work.

Because _Concurrency now depends on WinSDK, a load of tests implicitly
depend on it, and that means that we need to make sure the tests get the
correct flags to make it work.

rdar://137111020

al45tair commented:
@swift-ci Please smoke test

compnerd commented on Runtimes/Core/Concurrency/CMakeLists.txt:
```diff
@@ -125,6 +126,7 @@ target_link_libraries(swift_Concurrency PRIVATE
   swiftConcurrencyInternalShims
   $<$<BOOL:${BUILD_SHARED_LIBS}>:swiftThreading>
   $<$<PLATFORM_ID:Windows>:Synchronization>
+  $<$<PLATFORM_ID:Windows>:WinSDK>
```
CC: @etcwilde - I'm not sure how this works. This is an upwards dependency on the Platform Overlay which is dependent on core. This is a circular dependency now.

al45tair replied:
If the desire is that e.g. `WinSDK` should be able to use `_Concurrency`, then we should split the former into e.g. `WinSDKCore` and `WinSDK`, with `WinSDKCore` usable from here and `WinSDK` importing and re-exporting `WinSDKCore` as well as any Concurrency-related extensions.

al45tair replied:
As for how it works currently, `WinSDK.swift` isn't using `_Concurrency`, and implicit `_Concurrency` import is disabled in the overlays in the new build system.

compnerd replied:
It is the former - we cannot use `WinSDK` from `_Concurrency` because it is part of the "Core" project which is built before the overlays are ready because the overlays are dependent on "Core".

For what you are trying to do - use `WinSDK` from `_Concurrency`, we should be able to accommodate that by moving `_Concurrency` from "Core" to "Supplemental".

compnerd replied:
(FWIW, it is just more stuff that will need to be fixed, similar to the tests)

al45tair replied:
Another possibility is just to create a whole new repository for these and have them as a separate Swift package.

compnerd commented on test/lit.cfg:
```diff
@@ -645,15 +645,33 @@ else:
     )
     toolchain_lib_dir = make_path(config.swift_lib_dir, 'swift', host_os)
+    if kIsWindows:
+      windows_extras = ' '.join([
```
Why? `' '.join([])` is just the empty string, why compute that?

compnerd commented on utils/build.ps1:
```diff
@@ -1361,6 +1361,9 @@ function Build-CMakeProject {
             } else {
+              # If you update this, you also need to update test/lit.cfg, which
+              # adds these arguments to config.host_build_swift as well as
+              # putting them into config.*_overlay_opt.
```
I wonder if we can embed these into the build system and entirely avoid the addition here.

al45tair replied:
That would be nice, but I don't think that's a change I want to make as part of this PR; I'm just trying to unblock merging this PR.

al45tair commented:
@swift-ci please smoke test macOS platform
